test/tools/test_tools_memories.py

```diff
@@ -58,3 +62,412 @@ def test_sopel_identifier_memory_channel_str():
+
+    memory.update(tuple_)
+    assert len(memory) == 2
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is True
+    assert 'froMlisTpairS' not in memory
+    assert 'froMdicT' not in memory
+    assert 'froMseT' not in memory
+
+    memory.update(list_)
+    assert len(memory) == 3
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is False
+    assert memory['froMlisTpairS'] is True
+    assert 'froMdicT' not in memory
+    assert 'froMseT' not in memory
+
+    memory.update(dict_)
+    assert len(memory) == 4
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is False
+    assert memory['froMlisTpairS'] is False
+    assert memory['froMdicT'] is True
+    assert 'froMseT' not in memory
+
+    memory.update(set_)
+    assert len(memory) == 5
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is False
+    assert memory['froMlisTpairS'] is False
+    assert memory['froMdicT'] is False
+    assert memory['froMseT'] is True
+
+
+# TODO: remove condition when dropping Python 3.8
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="'|' operator added in py3.9")
+def test_sopel_identifier_memory_or_op():
+    memory = memories.SopelIdentifierMemory({'FromMemory': True})
+    assert len(memory) == 1
+    assert memory['FromMemory'] is True
+    assert memory['frommemory'] is True
+    assert memory['froMmemorY'] is True
+    assert 'FromDict' not in memory
+
+    other_memory = memories.SopelIdentifierMemory({
+        'FromMemory': False,
+        'FromOther': True,
+    })
+    assert len(other_memory) == 2
+    assert 'FromOther' in other_memory
+    assert other_memory['FromOther'] is True
+    assert 'fromother' in other_memory
+
+    result = memory | other_memory
+    assert len(result) == 2
+    assert result['fROmMemorY'] is False
+    assert result['fROmoTHEr'] is True
+    assert 'FromTuple' not in result
+
+
+# TODO: remove condition when dropping Python 3.8
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="'|' operator added in py3.9")
+def test_sopel_identifier_memory_or_op_dict():
+    memory = memories.SopelIdentifierMemory({'FromMemory': True, 'FromDict': False})
+    assert len(memory) == 2
+    assert memory['FromMemory'] is True
+    assert memory['frommemory'] is True
+    assert memory['froMmemorY'] is True
+    assert memory['FromDict'] is False
+
+    dictionary = {'FromMemory': False, 'FromDict': True}
+    assert len(dictionary) == 2
+    assert dictionary['FromDict'] is True
+    assert dictionary['FromMemory'] is False
+
+    # __or__
+    result = memory | dictionary
+    assert len(result) == 2
+    assert result['fROmMemorY'] is False
+    assert result['fROmDicT'] is True
+
+    # __ror__
+    result = dictionary | memory
+    assert len(result) == 2
+    assert result['fROmMemorY'] is True
+    assert result['fROmDicT'] is False
```

Comment:
It feels like these tests have a decent amount of repetition in them to cover case inflections, and would be a lot more compact and to-the-point if that functionality was tested elsewhere. As I've gone through this PR I've become more convinced that a test (or series of tests) dedicated to case would help the legibility of the rest of the added tests.

Reply:
Admittedly, we already have tests of case inflections for both `str` and `Identifier` keys; take a look near the top of the file. I even got pretty dang tired of coming up with alternative versions of each key while writing the tests, but it felt necessary.

Many of the behaviors this patch touches needed fixing because they broke handling of case inflections in one way or another (letting non-`Identifier` keys into the object; returning a `dict` instead of `SopelIdentifierMemory`; etc.). The test cases check multiple inflections as a defense-in-depth against breaking the key-casting behavior again.

I'm also not convinced that what's here is enough in all cases. When you asked about using `self.copy()` in `__ror__()` above, I changed it before thinking, ran the test suite, and it still passed. (Next push will include a modification to catch this in at least one of the `or_op` test cases.)

Reply:
While reading on a first pass, I had some thoughts about how we might try to push the case inflection testing into parameters or fixtures, I will mull that over a bit more for whenever I come back to this.

Reply:
Any results of said mulling to address in this PR, or should we resolve the thread? I ask because you did approve the changes after all, so maybe the result was "it's not worth the refactoring effort".